cpp/open3d/t/pipelines/slac/SLACOptimizer.cpp

```diff
@@ -46,56 +46,170 @@ namespace slac {
+
+    // Only take valid indices.
+    core::Tensor target_indices =
+            target_correspondences.IndexGet({valid_correspondences});
+
+    // Number of good correspondences (C).
+    int64_t C = target_indices.GetLength();
+
+    // correpondence_set : (i, corres[i]).
+    // source[i] and target[corres[i]] is a correspondence.
+    core::Tensor source_indices =
+            core::Tensor::Arange(0, N, 1, core::Dtype::Int64, device)
+                    .IndexGet({valid_correspondences})
+                    .Reshape({C, 1});
+
+    // Creating {C, 2} shaped tensor by horizontal stacking {source_indices,
+    // target_indices}.
+    core::Tensor correspondence_set({C, 2}, core::Dtype::Int64, device);
+    correspondence_set.SetItem(
+            {core::TensorKey::Slice(0, C, 1), core::TensorKey::Slice(0, 1, 1)},
+            source_indices);
+    correspondence_set.SetItem(
+            {core::TensorKey::Slice(0, C, 1), core::TensorKey::Slice(1, 2, 1)},
+            target_indices);
+
+    return correspondence_set;
 }
 
-/// Aggressive pruning -- reject any suspicious pair
-core::Tensor GetCorrespondencesForPair(int i,
-                                       int j,
-                                       TPointCloud& tpcd_i,
-                                       TPointCloud& tpcd_j,
-                                       const core::Tensor& T_i,
-                                       const core::Tensor& T_j,
-                                       const core::Tensor& T_ij,
-                                       float threshold) {
-    // Obtain correspondence via nns
-    auto result = open3d::pipelines::registration::EvaluateRegistration(
-            tpcd_i.ToLegacyPointCloud(), tpcd_j.ToLegacyPointCloud(), threshold,
-            core::eigen_converter::TensorToEigenMatrixXf(T_ij).cast<double>());
-    core::Tensor corres = core::eigen_converter::EigenVector2iVectorToTensor(
-            result.correspondence_set_, core::Dtype::Int64,
-            core::Device("CPU:0"));
-
-    TPointCloud tpcd_i_cropped(tpcd_i.GetPoints().IndexGet({corres.T()[0]}));
-    TPointCloud tpcd_j_cropped(tpcd_j.GetPoints().IndexGet({corres.T()[1]}));
-    tpcd_i_cropped.Transform(T_i);
-    tpcd_j_cropped.Transform(T_j);
-
-    // N x 3
-    core::Tensor residual =
-            tpcd_i_cropped.GetPoints() - tpcd_j_cropped.GetPoints();
-    core::Tensor square_residual = (residual * residual).Sum({1});
-    core::Tensor inliers = square_residual.Le(threshold * threshold);
+// Aggressive pruning -- reject any suspicious pair
+//
+// tpcd_i is the source pointcloud, tpcd_j is the target pointcloud,
+// T_i is the transformation_model_to_source,
+// T_j is the transformation_model_to_target,
+// T_ij is the transformation_source_to_target.
+// distance_threshold is the search_distance for NNS.
+// i and j are the indices of source and target pointcloud respectively.
+static core::Tensor GetCorrespondenceSetForPointCloudPair(
```

Comment:
`Correspondence` related functions are currently not exposed, as there are a few points to discuss related to its design, after the release. It will be part of the `registration refactoring`. So, for now, I have kept them mostly local to this file. If you want, I can expose them too.